Language/Haskell/GhcMod/ImportedFrom.hs

```diff
@@ -0,0 +1,812 @@
+    nameCacheFromGhc :: forall m. (GhcMonad m, MonadIO m) => Haddock.NameCacheAccessor m
+    nameCacheFromGhc = ( read_from_session , write_to_session )
+      where
+        read_from_session = do
+           ref <- GhcMonad.withSession (return . hsc_NC)
+           liftIO $ readIORef ref
+        write_to_session nc' = do
+           ref <- GhcMonad.withSession (return . hsc_NC)
+           liftIO $ writeIORef ref nc'
+
+getModuleExports
+    :: forall m. (GhcMonad m, GmOut m, GmLog m, IOish m)
+    => FilePath
+    -> (FilePath -> [String] -> String -> IO String)
+    -> [GhcPkgDb]
+    -> NiceImportDecl
+    -> m (Maybe ([String], String))
+getModuleExports ghcPkg readProc pkgDbStack m = do
+    minfo     <- (findModule (mkModuleName $ modName m) Nothing >>= getModuleInfo)
+                   `gcatch` (\(e :: SourceError) -> do gmLog GmDebug "getModuleExports" $ strDoc
+                                                        $ "Failed to find module \"" ++ modName m
+                                                        ++ "\": " ++ show e
+                                                       return Nothing)
+
+    p <- pkgFindModule ghcPkg readProc pkgDbStack (modName m)
+
+    dflags <- GHC.getSessionDynFlags
+
+    case (minfo, p) of
+        (Nothing,      _) -> return Nothing
+        (_,           "") -> return Nothing
+        (Just minfo',  _) -> return $ Just (map (showSDocForUser dflags ghcQualify . ppr) $ modInfoExports minfo', p)
+
+type FullyQualifiedName = String    -- ^ e.g. e.g. "base-4.8.2.0:Data.Foldable.length"
+type StrModuleName      = String    -- ^ e.g. "Data.List"
+
+data MySymbol = MySymbolSysQualified  String  -- ^ e.g. "base-4.8.2.0:Data.Foldable.length"
+              | MySymbolUserQualified String  -- ^ e.g. "DL.length" with an import earlier like "import qualified Data.List as DL"
+              deriving Show
+
+data ModuleExports = ModuleExports
+    { mName            :: StrModuleName            -- ^ e.g. "Data.List"
+    , mPackageName     :: String                   -- ^ e.g. "snap-0.14.0.6"
+    , mInfo            :: NiceImportDecl           -- ^ Our parse of the module import, with info like "hiding (map)".
+    , qualifiedExports :: [FullyQualifiedName]     -- ^ e.g. [ "base-4.8.2.0:GHC.Base.++"
+                                                   --        , "base-4.8.2.0:GHC.List.filter"
+                                                   --        , "base-4.8.2.0:GHC.List.zip"
+                                                   --        , ...
+                                                   --        ]
+    }
+    deriving Show
+
+refineAs :: forall m. MonadError GhcModError m => MySymbol -> [ModuleExports] -> m [ModuleExports]
+
+-- User qualified the symbol, so we can filter out anything that doesn't have a matching 'modImportedAs'.
+refineAs (MySymbolUserQualified userQualSym) exports = filterM f exports
+  where
+    -- f :: ModuleExports -> Bool
+    f export = case modas of
```

Comment:
I replaced this entire function with `modAs == moduleOfQualifiedName userQualSym`. Don't you think a datatype called MySymbolUserQualified should maintain the invariant that the module it contains is actually qualified ;)

The way we deal with this usually is by using smart constructors (`mkSymbolUserQualified` etc.) that check this invariant and only use those to construct that type. (optionally moving the type definition to a separate module that doesn't export the constructor so the mkFun in the only way to make a value of that type.